File: crates/flowrt-codegen/tests/golden/service_rust/expected/rust/tests/message_abi.rs
// FlowRT 管理产物。不要手工修改。

fn bytes_of<T>(value: &T) -> Vec<u8> {
    let mut bytes = vec![0u8; std::mem::size_of::<T>()];
    // Safety：生成测试只传入 FlowRT ABI v0.1 plain-data 消息，且 padding 已初始化。
    unsafe {
        std::ptr::copy_nonoverlapping(
            (value as *const T).cast::<u8>(),
            bytes.as_mut_ptr(),
            bytes.len(),
        );
    }
    bytes
}

fn assert_default_bytes_zero<T: Copy + Default>() {
    let value = T::default();
    assert_eq!(bytes_of(&value), vec![0u8; std::mem::size_of::<T>()]);
}

fn assert_byte_roundtrip<T: Copy + Default>(value: T) {
    let bytes = bytes_of(&value);
    let mut roundtrip = T::default();
    // Safety：`roundtrip` 是有效 plain-data 存储，`bytes` 长度等于 `size_of::<T>()`。
    unsafe {
        std::ptr::copy_nonoverlapping(
            bytes.as_ptr(),
            (&mut roundtrip as *mut T).cast::<u8>(),
            bytes.len(),
        );
    }
    assert_eq!(bytes_of(&roundtrip), bytes);
}

fn assert_sample_bytes<T: Copy>(value: T, expected: &[u8]) {
    assert_eq!(bytes_of(&value), expected);
}

const EXPECTED_PLAN_REQUEST_BYTES: &[u8] = &[2, 0, 0, 0];
const EXPECTED_PLAN_RESPONSE_BYTES: &[u8] = &[1];

fn sample_plan_request() -> flowrt_app::messages::PlanRequest {
    let mut value = flowrt_app::messages::PlanRequest::default();
    value.goal = 2u32;
    value
}

fn sample_plan_response() -> flowrt_app::messages::PlanResponse {
    let mut value = flowrt_app::messages::PlanResponse::default();
    value.accepted = true;
    value
}

#[test]
fn plan_request_message_abi() {
    assert_eq!(std::mem::size_of::<flowrt_app::messages::PlanRequest>(), 4);
    assert_eq!(std::mem::align_of::<flowrt_app::messages::PlanRequest>(), 4);
    assert_default_bytes_zero::<flowrt_app::messages::PlanRequest>();
    assert_eq!(std::mem::offset_of!(flowrt_app::messages::PlanRequest, goal), 0);
    assert_byte_roundtrip(sample_plan_request());
    assert_sample_bytes(sample_plan_request(), EXPECTED_PLAN_REQUEST_BYTES);
}

#[test]
fn plan_response_message_abi() {
    assert_eq!(std::mem::size_of::<flowrt_app::messages::PlanResponse>(), 1);
    assert_eq!(std::mem::align_of::<flowrt_app::messages::PlanResponse>(), 1);
    assert_default_bytes_zero::<flowrt_app::messages::PlanResponse>();
    assert_eq!(std::mem::offset_of!(flowrt_app::messages::PlanResponse, accepted), 0);
    assert_byte_roundtrip(sample_plan_response());
    assert_sample_bytes(sample_plan_response(), EXPECTED_PLAN_RESPONSE_BYTES);
}

File: crates/flowrt-codegen/tests/golden/iox2_service_cpp/expected/cpp/tests/message_abi.cpp
// FlowRT 管理产物。不要手工修改。

#include <array>
#include <cassert>
#include <cstddef>
#include <cstdint>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <type_traits>

#include "flowrt_app/messages.hpp"

namespace {

template <typename T>
std::array<std::uint8_t, sizeof(T)> bytes_of(const T& value) {
    std::array<std::uint8_t, sizeof(T)> bytes{};
    std::memcpy(bytes.data(), &value, bytes.size());
    return bytes;
}

template <typename T>
void assert_default_bytes_zero() {
    T value{};
    std::array<std::uint8_t, sizeof(T)> expected{};
    assert(bytes_of(value) == expected);
}

template <typename T>
void assert_byte_roundtrip(const T& value) {
    const auto bytes = bytes_of(value);
    T roundtrip{};
    std::memset(&roundtrip, 0, sizeof(roundtrip));
    std::memcpy(&roundtrip, bytes.data(), bytes.size());
    assert(std::memcmp(&roundtrip, &value, sizeof(T)) == 0);
}

template <typename T, std::size_t N>
void assert_sample_bytes(const T& value, const std::array<std::uint8_t, N>& expected) {
    static_assert(sizeof(T) == N);
    assert(bytes_of(value) == expected);
}

template <std::size_t N>
void write_fixture(std::string_view name, const std::array<std::uint8_t, N>& bytes) {
#ifdef FLOWRT_ABI_FIXTURE_DIR
    std::filesystem::create_directories(FLOWRT_ABI_FIXTURE_DIR);
    auto path = std::filesystem::path(FLOWRT_ABI_FIXTURE_DIR) / std::string(name);
    std::ofstream output(path, std::ios::binary);
    if (!output) {
        throw std::runtime_error("failed to open ABI fixture output");
    }
    output.write(reinterpret_cast<const char*>(bytes.data()), static_cast<std::streamsize>(bytes.size()));
    if (!output) {
        throw std::runtime_error("failed to write ABI fixture output");
    }
#else
    (void)name;
    (void)bytes;
#endif
}

constexpr std::array<std::uint8_t, 4> EXPECTED_PLAN_REQUEST_BYTES{{2, 0, 0, 0}};
constexpr std::array<std::uint8_t, 1> EXPECTED_PLAN_RESPONSE_BYTES{{1}};

flowrt_app::PlanRequest sample_plan_request() {
    flowrt_app::PlanRequest value{};
    std::memset(&value, 0, sizeof(value));
    value.goal = std::uint32_t{2};
    return value;
}

flowrt_app::PlanResponse sample_plan_response() {
    flowrt_app::PlanResponse value{};
    std::memset(&value, 0, sizeof(value));
    value.accepted = true;
    return value;
}

void test_plan_request_message_abi() {
    static_assert(std::is_standard_layout_v<flowrt_app::PlanRequest>);
    static_assert(std::is_trivially_copyable_v<flowrt_app::PlanRequest>);
    static_assert(sizeof(flowrt_app::PlanRequest) == 4);
    static_assert(alignof(flowrt_app::PlanRequest) == 4);
    assert_default_bytes_zero<flowrt_app::PlanRequest>();
    static_assert(offsetof(flowrt_app::PlanRequest, goal) == 0);
    assert_byte_roundtrip(sample_plan_request());
    assert_sample_bytes(sample_plan_request(), EXPECTED_PLAN_REQUEST_BYTES);
    write_fixture("plan_request.bin", bytes_of(sample_plan_request()));
}

void test_plan_response_message_abi() {
    static_assert(std::is_standard_layout_v<flowrt_app::PlanResponse>);
    static_assert(std::is_trivially_copyable_v<flowrt_app::PlanResponse>);
    static_assert(sizeof(flowrt_app::PlanResponse) == 1);
    static_assert(alignof(flowrt_app::PlanResponse) == 1);
    assert_default_bytes_zero<flowrt_app::PlanResponse>();
    static_assert(offsetof(flowrt_app::PlanResponse, accepted) == 0);
    assert_byte_roundtrip(sample_plan_response());
    assert_sample_bytes(sample_plan_response(), EXPECTED_PLAN_RESPONSE_BYTES);
    write_fixture("plan_response.bin", bytes_of(sample_plan_response()));
}

}  // namespace

int main() {
    test_plan_request_message_abi();
    test_plan_response_message_abi();
    return 0;
}
